pkg/controllers/nodeclaim/capacityreservation/controller.go

```diff
@@ -0,0 +1,153 @@
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if updated {
+			updatedNodeClaims.Insert(ncs.Items[i].Name)
+		}
+	}
+	log.FromContext(ctx).WithValues("NodeClaims", lo.Map(updatedNodeClaims.UnsortedList(), func(name string, _ int) klog.ObjectRef {
+		return klog.KRef("", name)
+	})).V(1).Info("updated capacity type for nodeclaims")
+
+	if len(errs) != 0 {
+		if lo.EveryBy(errs, func(err error) bool { return errors.IsConflict(err) }) {
+			return reconcile.Result{Requeue: true}, nil
+		}
+		return reconcile.Result{}, multierr.Combine(errs...)
+	}
+	return reconcile.Result{RequeueAfter: time.Minute}, nil
+}
+
+// syncCapacityType will update the capacity type for the given NodeClaim. This accounts for the fact that capacity
+// reservations will expire, demoting NodeClaims with capacity type "reserved" to "on-demand".
+func (c *Controller) syncCapacityType(ctx context.Context, capacityType string, nc *karpv1.NodeClaim) (bool, error) {
+	// We won't be able to sync deleting NodeClaims, and there's no real need to either as they're already draining.
+	if !nc.DeletionTimestamp.IsZero() {
+		return false, nil
+	}
+
+	// For now we only account for the case where a reserved NodeClaim becomes an on-demand NodeClaim. This does not
+	// account for on-demand NodeClaims being promoted to reserved since that is not natively supported by Karpenter.
+	if capacityType != karpv1.CapacityTypeOnDemand {
+		return false, nil
+	}
+	if nc.Labels[karpv1.CapacityTypeLabelKey] == karpv1.CapacityTypeReserved {
+		stored := nc.DeepCopy()
+		nc.Labels[karpv1.CapacityTypeLabelKey] = karpv1.CapacityTypeOnDemand
+		delete(nc.Labels, cloudprovider.ReservationIDLabel)
+		if err := c.kubeClient.Patch(ctx, nc, client.MergeFrom(stored)); err != nil {
+			return false, fmt.Errorf("patching nodeclaim %q, %w", nc.Name, err)
+		}
+	}
+
+	// If the reservation expired before the NodeClaim became registered, there may not be a Node on the cluster. Note
+	// that there should never be duplicate Nodes for a given NodeClaim, but handling this user-induced error is more
+	// straightforward than handling the duplicate error.
+	nodes, err := nodeclaimutils.AllNodesForNodeClaim(ctx, c.kubeClient, nc)
+	if err != nil {
+		return false, fmt.Errorf("listing nodes for nodeclaim %q, %w", nc.Name, err)
+	}
+	for _, n := range nodes {
+		if !n.DeletionTimestamp.IsZero() {
+			continue
+		}
+		// Skip Nodes which haven't been registered since we still may not have synced labels. We'll get it on the next
+		// iteration.
+		if n.Labels[karpv1.NodeRegisteredLabelKey] != "true" {
+			continue
+		}
+		if n.Labels[karpv1.CapacityTypeLabelKey] != karpv1.CapacityTypeReserved {
+			continue
+		}
+		stored := n.DeepCopy()
+		n.Labels[karpv1.CapacityTypeLabelKey] = karpv1.CapacityTypeOnDemand
+		delete(nc.Labels, cloudprovider.ReservationIDLabel)
+		if err := c.kubeClient.Patch(ctx, n, client.MergeFrom(stored)); err != nil {
+			return false, fmt.Errorf("patching node %q, %w", n.Name, err)
```

Comment:
What about the node NotFound error?

Reply:
Ditto with the nodeclaim not found, we should ignore